boostorg/winapi/issues/70 Support Windows CE 6.00, 7.00 and WM 6.x

xentrax opened the pull request:
As discussed in https://github.com/boostorg/winapi/issues/70 introduced use of 
BOOST_WINAPI_IMPORT and BOOST_WINAPI_IMPORT_EXCEPT_WM instead of BOOST_SYMBOL_IMPORT.

The macros were declared in config.hpp

Also fixed issues caused by different signatures of some WINAPI calls in 
critical_section.hpp
process.hpp
local_memory.hpp

Lastique commented on include/boost/winapi/process.hpp:
```diff
@@ -428,7 +455,11 @@ BOOST_FORCEINLINE BOOL_ CreateProcessW(
+#if defined(_WIN32_WCE)
+    LPWSTR_ lpCurrentDirectory,
+#else
+	LPCWSTR_ lpCurrentDirectory,
```
No tabs, please. Here and everywhere else.

xentrax replied:
agree

Lastique commented on include/boost/winapi/process.hpp:
```diff
@@ -47,15 +47,18 @@ struct _STARTUPINFOEXW;
 
-BOOST_SYMBOL_IMPORT BOOST_NORETURN boost::winapi::VOID_ BOOST_WINAPI_WINAPI_CC
+#if !defined(_WIN32_WCE)
+// On Windows CE ExitProcess is a macro to call TerminateProcess
```
In this case, Boost.WinAPI has to provide a portable way to call `ExitProcess`, including in cases when it is a macro and when it is not defined. The best approach is to provide an inline wrapper function in `boost::winapi` (e.g. `exit_process`).

xentrax replied:
agree

Lastique commented on include/boost/winapi/process.hpp:
```diff
@@ -68,7 +71,20 @@ BOOST_SYMBOL_IMPORT boost::winapi::BOOL_ BOOST_WINAPI_WINAPI_CC CreateProcessA(
+	boost::winapi::INT_ bInheritHandles,
+	boost::winapi::DWORD_ dwCreationFlags,
+	boost::winapi::LPVOID_ lpEnvironment,
+	boost::winapi::LPWSTR_ lpCurrentDirectory,
```
Here and in other functions, changing `LPCWSTR_` to `LPWSTR_` is not portable as it breaks user's code which passes constant strings as arguments. If this function actually does not modify the argument, there needs to be an inline wrapper function in `boost::winapi` which will cast away the constness.

xentrax replied:
agree

xentrax commented on include/boost/winapi/critical_section.hpp:
```diff
@@ -20,89 +20,97 @@
+// Windows CE uses a different name for the structure
+#if defined (_WIN32_WCE)
+struct CRITICAL_SECTION;
+typedef CRITICAL_SECTION BOOST_WINAPI_RTL_CRITICAL_SECTION;
```
This suggestion, if implemented literally, causes 3 issues.
1.  public boost::winapi interface mentions detail namespace
2.  it create illegal overload with extern "C" linkage
3. it causes ambiguity if unqualified call (like error C2668: 'EnterCriticalSection': ambiguous call to overloaded function)

I can amend the commit to exclude the change about typedef for _RTL_CRITICAL_SECTION altogether, and open separate PR only about that. What do you think?

Lastique replied:
1. Not sure I understand what you mean. Public declarations in `boost::winapi` do not and should not use neither `::_RTL_CRITICAL_SECTION` nor `::CRITICAL_SECTION` but a dedicated structure `boost::winapi::_RTL_CRITICAL_SECTION`. The typedef I suggested (let's call it `boost::winapi::detail::winsdk_critical_section`) will only be used in the declarations in the global namespace - i.e. those mirroring the ones from Windows SDK.
2. Since the typedef will refer to the same structure that is defined in the Windows SDK, this will not be an overload but rather a declaration of the same function.
3. Unqualified calls from within Boost should not happen since all Boost.WinAPI public symbols are in `boost::winapi`. Any Boost library that is using Boost.WinAPI has to specify at the very least `winapi` namespace. Unqualified calls from non-Boost user's code will resolve to the function from the Windows SDK (see point 2; the declaration in Boost.WinAPI must be equivalent to that in Windows SDK and thus not introduce an overload).

xentrax replied:
ok, it works, sorry.

3 more questions.

1.
Do you prefer boost::winapi::detail::winapi_critical_section
or boost::winapi::detail::winsdk_critical_section ?

2. 
I can apply same approach for MINGW #ifdef in the same file for consistency.

3.
The internal layout of the CRITICAL_SECTION  on CE is different from NT. Fortunately, sizeof of CE variant is smaller.

boost::winapi::_RTL_CRITICAL_SECTION mimics NT. Do we want to do something about it?

This is CE variant:

`typedef` struct CRITICAL_SECTION {
    unsigned int LockCount;         /* Nesting count on critical section */
    HANDLE OwnerThread;         	/* Handle of owner thread */
    HANDLE hCrit;					/* Handle to this critical section */
    DWORD needtrap;					/* Trap in when freeing critical section */
    DWORD dwContentions;			/* Count of contentions */
} CRITICAL_SECTION, `*LPCRITICAL_SECTION;`

Lastique replied:
1. `winsdk_critical_section` is better as it implies that it refers to the type defined in Windows SDK rather than Boost.WinAPI.
2. Yes, that would be great, thanks. Do check that other than the structure name the function signatures are the same, though.
3. I think it would be better to mimic the CE structure on Windows CE.

xentrax commented:
The changes discussed so far are implemented in 3 commits because the 17fbd4a and 10c4126 are 
somewhat independent and, also, not critical for the original problem.

Is the anything else I can do here?

Lastique commented on include/boost/winapi/process.hpp:
```diff
@@ -366,7 +390,15 @@ using ::OpenProcess;
+BOOST_FORCEINLINE BOOST_NORETURN boost::winapi::VOID_ exit_process(
+    boost::winapi::UINT_ uExitCode)
+{
+    ExitProcess(uExitCode);
```
Where is this macro defined on CE? You haven't added any #includes to this header.

xentrax replied:
Good catch, thanks. To support compilation without including any headers I will call TerminateProcess directly instead of "calling" ExitProcess macro.

Lastique replied:
Yes, though no need for the additional parenthesis around `uExitCode` since we're not in a macro.

Lastique commented on include/boost/winapi/critical_section.hpp:
```diff
@@ -20,90 +20,122 @@
+namespace boost {
+namespace winapi {
+namespace detail {
+    typedef CRITICAL_SECTION winapi_critical_section;
```
I thought we settled on `winsdk_critical_section`, didn't we?

Lastique commented on include/boost/winapi/critical_section.hpp:
```diff
@@ -125,6 +160,18 @@ typedef struct BOOST_MAY_ALIAS _RTL_CRITICAL_SECTION {
 
+#else
+
+typedef struct BOOST_MAY_ALIAS CRITICAL_SECTION {
```
Better use the same name Windows SDK uses for the structure, i.e. `_CRITICAL_SECTION`.

xentrax replied:
Yes, big Windows SDK uses "_RTL_CRITICAL_SECTION" name in its SDK.
Windows CE SDKs uses CRITICAL_SECTION name.

I don't see any harm in using _RTL_CRITICAL_SECTION here, but IMHO it seems inconsistent with your previous strict approach, and may confuse other developers targeting Windows CE.

Of course, I will use _RTL_CRITICAL_SECTION. But if you have a second thought, no problem, just tell.

Lastique commented:
Also, your PR seems to include some changes that are in develop. Could you rebase the PR against the current develop? It would make reviewing a little easier.

Lastique commented:
> Yes, big Windows SDK uses "_RTL_CRITICAL_SECTION" name in its SDK.
> Windows CE SDKs uses CRITICAL_SECTION name.

Oh, I misread the code then (for some reason I thought CRITICAL_SECTION 
was a typedef to _CRITICAL_SECTION). Sure, let's leave CRITICAL_SECTION 
then.

xentrax commented:
I fixed the updated list of problems, I tried hard to not forget about winsdk_critical_section and tabs this time.

I reinstated CRITICAL_SECTION name to match Windows CE SDKs.
 
Inspired by the ExitProcess remark I also fixed a small compilation issue for the case when boost/winapi headers are included before anything else. (A couple of typedefs added in basic_types.hpp).
